cpp/mg_utility/mage.hpp

```diff
@@ -0,0 +1,2757 @@
+}
+
+inline Node Nodes::Iterator::operator*() {
+  if (nodes_iterator_ == nullptr) return Node((const mgp_vertex *)nullptr);
+
+  auto node = Node(mgp::vertices_iterator_get(nodes_iterator_));
+  return node;
+}
+
+inline Nodes::Iterator Nodes::begin() { return Iterator(nodes_iterator_); }
+
+inline Nodes::Iterator Nodes::end() { return Iterator(nullptr); }
+
+// GraphRelationships:
+
+inline GraphRelationships::Iterator::Iterator(mgp_vertices_iterator *nodes_iterator) : nodes_iterator_(nodes_iterator) {
+  // Positions the iterator over the first existing relationship
+
+  if (nodes_iterator_ == nullptr) {
+    return;
+  }
+
+  // Go through each graph node’s adjacent nodes
+  for (auto node = mgp::vertices_iterator_get(nodes_iterator_); node;
+       node = mgp::vertices_iterator_next(nodes_iterator_)) {
+    // Check if node exists
+    if (node == nullptr) {
+      mgp::vertices_iterator_destroy(nodes_iterator_);
+      nodes_iterator_ = nullptr;
+      return;
+    }
+
+    // Check if node has out-relationships
+    out_relationships_iterator_ = mgp::vertex_iter_out_edges(node, memory);
+    auto relationship = mgp::edges_iterator_get(out_relationships_iterator_);
+    if (relationship != nullptr) return;
+
+    mgp::edges_iterator_destroy(out_relationships_iterator_);
+    out_relationships_iterator_ = nullptr;
+  }
+}
+
+inline GraphRelationships::Iterator::~Iterator() {
+  if (nodes_iterator_ != nullptr) {
+    mgp::vertices_iterator_destroy(nodes_iterator_);
+  }
+  if (out_relationships_iterator_ != nullptr) {
+    mgp::edges_iterator_destroy(out_relationships_iterator_);
+  }
+}
+
+inline GraphRelationships::Iterator &GraphRelationships::Iterator::operator++() {
+  // Moves the iterator onto the next existing relationship
+
+  // 1. Check if the current node has remaining relationships to iterate over
+
+  if (out_relationships_iterator_ != nullptr) {
+    auto next = mgp::edges_iterator_next(out_relationships_iterator_);
+
+    if (next == nullptr) {
+      mgp::edges_iterator_destroy(out_relationships_iterator_);
+      out_relationships_iterator_ = nullptr;
+    }
+  }
+
+  // 2. Move onto the next nodes
+
+  if (nodes_iterator_ != nullptr) {
+    for (auto node = mgp::vertices_iterator_next(nodes_iterator_); node;
+         node = mgp::vertices_iterator_next(nodes_iterator_)) {
+      // Check if node exists - if it doesn’t, we’ve reached the end of the iterator
+      if (node == nullptr) {
+        mgp::vertices_iterator_destroy(nodes_iterator_);
+        nodes_iterator_ = nullptr;
+        return *this;
+      }
+
+      // Check if node has out-relationships
+      out_relationships_iterator_ = mgp::vertex_iter_out_edges(node, memory);
+      auto relationship = mgp::edges_iterator_get(out_relationships_iterator_);
+      if (relationship != nullptr) {
+        return *this;
+      } else {
+        mgp::edges_iterator_destroy(out_relationships_iterator_);
+        out_relationships_iterator_ = nullptr;
+      }
```

Comment:
No need for else if return
```suggestion
      if (relationship != nullptr) {
        return *this;
      }
      mgp::edges_iterator_destroy(out_relationships_iterator_);
      out_relationships_iterator_ = nullptr;
```

Reply:
Done